### PR TITLE
Fix generated-ID row mismatches in PrepAndExpectedTestCase

### LINKED ISSUE #672: PrepAndExpectedTestCase should only sort on filtered columns
> **Migrated from SourceForge** — [Bugs #430](https://sourceforge.net/p/dbunit/bugs/430/)
> Originally reported by **lufecir** on 2019-04-11 12:56:05.283000
> Priority: `5`
> Assigned to **jeffjensen**
> SF status: pending

Hi,

Not sure if I'm missing something obvious, but it seems to me that PrepAndExpectedTestCase should only sort on the filtered columns, and not on all the columns.
Otherwise you might end up in a situation where testcases fail randomly (which actually happened for me). 
Suppose the table for which you are verifying expected results  contains a column with some randomly generated value as one of its first columns (e.g. aff_id). 
You obviously want to ignore this column, so it is in the list of excluded columns and not in your expected dataset. 
But if the datasets are sorted on all columns, then the actual dataset will be sorted on that excluded column, while expected results will not be, giving a potential (random) mismatch depending on the values of the 'random' column 'aff_id'.

Example:
Actual
aff_id | other_id | value
---------- | ---------- | ------ 
-315      | 25       | codeX
48      | 14       | codeY
Expected
other_id | value
---------- | ------ 
14       | codeY
25       | codeX

In the above example the verification will fail although - given 'aff_id' should be excluded - the datasets should be considered identitical. 

I have tested with reversing the order of the logic in DefaultPrepAndExpectedTestCase#verifyData and filtering the columns first before sorting the tables. This seems to work fine for me, but I'm just wondering if I'm not missing something else.
If not, I could create a merge request containing the fix. 

Regards,

Luc

---

**jeffjensen** — 2019-04-19 14:04:05.734000
Hi Luc, thank you for the discussion on this.

The current behavior of sorting on all fields defined in the dataset has existed since its creation.  So changing it breaks backwards compatability (many users of dbUnit would have failed tests due to sorting on the first included column).  It may not be desired this way tho, and can consider for change.

The solution for the example you gave is to add an id column as first one in the expected dataset.  dbUnit then ignores it on comparison and uses it for sorting.  This behavior helps having datasets in same order for comparison, which is a fundamental requirement.  Non-excluded columns may not have unique values to sort by.

Ignoring columns is one of the key reasons for the ValueComparer feature creation - so we don't have to ignore columns - enabling comparing ID fields with comparisons such as "actual greater than or equal to expected".

**lufecir** — 2019-05-02 11:17:42.534000
Hi Jeff,
Thanks for your reply. 
My problem is that the 'expected' datasets are being read from Excel files from a legacy testing framework. And we should still be able to run the legacy tests, so adding a column to the expected dataset could be causing problems (will have to investigate). Also there are a lot of those legacy files, so - even if adding a column would be possible - I want avoid making changes to them, since this will result in a lot of work. 

I understand that you cannot break existing functionality, so I will have to do some more research on how I can fix my issue, maybe by using the ValueComparer feature.

**jeffjensen** — 2019-05-02 22:40:00.584000
You have a good scenario, thank you for sharing.

A solution idea is to allow the user to decide - add a feature toggle for performing this sort or not; default is on so is backwards compatible.  You can turn it off.  WDYT?

**lufecir** — 2019-05-03 04:57:43.490000
Yes, that would be perfect for my use case. I was kind of reluctant to propose this, since at first it felt somehow like patchwork. 
But giving it some second thought, it could indeed be presented to the user as a proper feature that is easy to understand with proper documentation. 
I will create a Merge Request.

**jeffjensen** — 2019-05-03 11:42:36.498000
Sounds good, thank you Luc.

Please create a feature ticket for that work and we can close this one.

### LINKED ISSUE #676: Allow PrepAndExpectedTestCase to sort only on filtered columns instead of all columns
> **Migrated from SourceForge** — [Feature Requests #233](https://sourceforge.net/p/dbunit/feature-requests/233/)
> Originally reported by **lufecir** on 2019-07-04 09:56:19.841000
> Priority: `5`
> Assigned to *(unassigned)*
> SF status: open

In some cases it is useful (and desired) to have dbUnit not sort on all columns, but only on the filtered columns when retrieving data from the database to compare with the expected results. 

Therefore I would like to propose to add a feature to DbUnit that allows toggling how the data will be sorted for comparison: on all columns or on filtered columns only. 

See https://sourceforge.net/p/dbunit/bugs/430/ for more details on the reasoning behind this.

---

**lufecir** — 2019-07-04 09:59:04.624000
Forgot to mention, but cannot edit feature request anymore:
* Default behavior would be to sort on all columns, in order to guarantee full backwards compatibility with current approach.

### PR DESCRIPTION
**Pull request by jeffjensen**
## Summary
* `DefaultPrepAndExpectedTestCase#verifyData` always sorted the actual table by all of its native columns (generated/identity column included) but the expected table by only the columns its file declares, which typically omits that column since its value is unknown ahead of time — `excludeColumns`/`includeColumns` were applied to the comparison but never to the sort. When production code doesn't guarantee row insertion order (e.g. Hibernate reordering a batch insert), this misaligns same-data rows and fails the comparison with a false `DbComparisonFailure`.
* Add `VerifyTableDefinition#sortOnFilteredColumnsOnly` (default `false`, preserving prior behavior), the opt-in toggle proposed in #676: when `true`, both tables sort by only their `excludeColumns`/`includeColumns`-filtered columns instead of all native columns.
* Add `DefaultPrepAndExpectedTestCaseGeneratedIdRowOrderIT` reproducing the defect and proving the fix (default still reproduces it, opt-in fixes it, opt-in still catches genuine mismatches), using the existing `IDENTITY_TABLE` fixture; added that DDL to derby/mysql/postgresql/oracle so it runs on all 9 database profiles.
* Documented `sortOnFilteredColumnsOnly` in Javadoc and a new site "Sort Mode" section, cross-linked from the row-ordering guidance in `PrepAndExpectedTestCase.adoc`, `equality.adoc`, and `decorators.adoc`.

Fixes #672
Closes #676

## Test plan
- [x] `./mvnw clean test` — 1935 unit tests pass
- [x] `./mvnw clean verify -Phsqldb-2-7` — 351 ITs pass
- [x] `./mvnw clean verify -Ph2-1-4` — 351 ITs pass
- [x] `./mvnw clean verify -Pderby-10-14` — 351 ITs pass
- [x] `./mvnw clean install site` — no new Asciidoctor/Javadoc warnings
- [ ] CI matrix across all 9 databases (mysql/postgresql/oracle/mssql/db2 not run locally this session — new IDENTITY_TABLE DDL for those 4 uses standard, well-known syntax but wasn't executed live)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AebUvmVvD9HpqAxDnKEBqK

## Summary by Sourcery

Add an opt-in mode to DefaultPrepAndExpectedTestCase to sort tables on filtered columns only, preventing false comparison failures when the first column is a generated/identity value, and document and test this behavior across supported databases.

New Features:
- Introduce VerifyTableDefinition#sortOnFilteredColumnsOnly to control whether table comparison sorts on filtered columns only or all native columns.
- Add integration tests and XML datasets demonstrating generated-ID row order behavior and the new sort-on-filtered-columns-only option.

Bug Fixes:
- Fix false DbComparisonFailure results in DefaultPrepAndExpectedTestCase when tables have a generated/identity first column excluded from comparison, but sorting still used all native columns.

Enhancements:
- Refine DefaultPrepAndExpectedTestCase#verifyData to support the new sort-on-filtered-columns-only behavior while preserving backward-compatible defaults.
- Update release notes and documentation to describe the new sort mode and its impact on row ordering and identity columns.

Documentation:
- Document the sortOnFilteredColumnsOnly option in Javadoc and Asciidoc, including guidance on row ordering and identity/generated columns.

Tests:
- Add DefaultPrepAndExpectedTestCaseGeneratedIdRowOrderIT to validate default behavior, the new filtered-columns sort mode, and that genuine mismatches are still detected.
- Extend database-specific SQL setup (derby, mysql, postgresql, oracle) with an IDENTITY_TABLE fixture used by the new integration test.

## Summary by CodeRabbit

- **New Features**
  - Added an opt-in setting to sort table comparisons using only included columns.
  - Improves reliability when generated or identity columns are excluded and rows are inserted in different orders.
  - Existing behavior remains unchanged by default.

- **Documentation**
  - Added configuration guidance, sorting behavior details, and usage examples.

- **Tests**
  - Added coverage for generated-ID row ordering, matching data, and genuine mismatches across supported databases.

### CODE REVIEW

File: src/main/java/org/dbunit/DefaultPrepAndExpectedTestCase.java [hunk 1]
```diff
@@ -707,8 +707,12 @@ protected void verifyData(final IDatabaseConnection connection,
         final ITable actualTable =
                 loadTableDataFromDatabase(tableName, connection);
 
+        final boolean sortOnFilteredColumnsOnly =
+                verifyTableDefinition.isSortOnFilteredColumnsOnly();
+
         verifyData(expectedTable, actualTable, excludeColumns, includeColumns,
-                defaultValueComparer, columnValueComparers);
+                defaultValueComparer, columnValueComparers,
+                sortOnFilteredColumnsOnly);
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/DefaultPrepAndExpectedTestCase.java]: Good catch, fixed in 98cda0f5: `verifyData(IDatabaseConnection, VerifyTableDefinition)` now calls the six-argument overload when `sortOnFilteredColumnsOnly` is false and only calls the seven-argument one when it's true, so a subclass overriding the six-argument overload is still invoked in the (default) common case.

File: src/site/asciidoc/components/verifytabledefinition.adoc [hunk 2]
```diff
@@ -62,6 +64,38 @@ with a custom `VerifyTableDefinitionVerifier`. Note the setter has no automatic
...
+[source,java]
+----
+VerifyTableDefinition table = new VerifyTableDefinition(
+        "IDENTITY_TABLE", new String[] {"IDENTITY_TABLE_ID"});
+table.setSortOnFilteredColumnsOnly(true);
+----
+
+Default is `false`, preserving the historical sort-by-all-columns behavior; opt in per
+table rather than assuming it everywhere, since a table with a normal (non-generated, or
+included) primary key sorts correctly either way.
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/site/asciidoc/components/verifytabledefinition.adoc]: Fair point, reworded in 98cda0f5: the safety condition isn't really about a column being "generated" per se, it's about whether every column needed to distinguish otherwise-identical rows is present on both tables. Updated the text to state that directly.

File: src/test/java/org/dbunit/DefaultPrepAndExpectedTestCaseGeneratedIdRowOrderIT.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        tc.preTest();
+
+        // Control case: sortOnFilteredColumnsOnly=true must not mask a
+        // genuine data mismatch (not just a row order difference).
+        assertThatThrownBy(() -> tc.postTest())
+                .as("Expected tc.postTest() to throw DbComparisonFailure for"
+                        + " genuinely mismatched data even with"
+                        + " sortOnFilteredColumnsOnly=true, but it didn't.")
+                .isInstanceOf(DbComparisonFailure.class);
+    }
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/DefaultPrepAndExpectedTestCaseGeneratedIdRowOrderIT.java:142]: Added in 98cda0f5: three new tests in DefaultPrepAndExpectedTestCaseTest calling the seven-argument verifyData(...) directly against mock tables (no database) - testVerifyData_withSortOnFilteredColumnsOnlyFalse_generatedIdOutOfDataOrder_throwsOnMismatch, ...True_...passesWhenEqual, and an all-columns-excluded edge case (also covers Sourcery's empty-sort-column-list question below).

File: src/test/resources/sql/mysql.sql [hunk 4]
```diff
@@ -61,3 +61,14 @@ CREATE TABLE EMPTY_MULTITYPE_TABLE
    NUMERIC_COL NUMERIC(38, 0),
    TIMESTAMP_COL TIMESTAMP NULL,
    VARBINARY_COL VARBINARY(254)) ENGINE = InnoDB;
+
+----------------------------------------------------------------------------
+- IDENTITY_TABLE
+----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/resources/sql/mysql.sql:67]: Verified against this codebase's actual DDL execution path and declining this one: DdlExecutor.readSqlFromFile() strips every line starting with "-" client-side before any SQL reaches the database driver (`if (!line.startsWith("-")) { sqlBuffer.append(line); }`), so these separator/comment lines never reach MySQL's own parser - its comment-syntax rules don't apply here. This `-----...-----` (no space) style is also already used identically for every other table in this same file and in every other dialect's DDL, predating this PR. Empirically, MySQL CI is green on this exact content (see the "ITs (MySQL)" checks). Changing just my new block would be inconsistent with the rest of the file for no behavioral benefit.

**Comment by jeffjensen** [comment]
Addressing Sourcery's two points from the initial review:

1. **makeSortColumns duplicating applyColumnFilters** — intentional, not accidental duplication. `applyColumnFilters` operates on `ITable` (returns a filtered `ITable` via `DefaultColumnFilter.includedColumnsTable`/`excludedColumnsTable`); `makeSortColumns` needs `Column[]` (what `SortedTable`'s constructor takes), a different input/output shape `SortedTable` requires. Rather than reimplementing name-matching, `makeSortColumns` calls `DefaultColumnFilter.accept()` directly — the same underlying matcher (including wildcard pattern support) both `applyColumnFilters` call sites use internally — so the two stay semantically unified at that shared primitive instead of duplicating matching logic. Restructuring `applyColumnFilters` itself to share more than that would mean reshaping a working, tested method for a cosmetic gain; happy to revisit if that primitive drifts in practice.

2. **Empty sort-column array when everything is excluded** — verified: `SortedTable`'s comparator loop is `for (i = 0; i < _sortColumns.length; i++)`, so a zero-length array just returns 0 immediately for every pair, i.e. `Arrays.sort` leaves rows in their original (stable) order — no exception. This is fine rather than surprising: the corresponding comparison (via the same exclude/include filters) also has zero columns to compare in that scenario, so row order doesn't matter for correctness either way. A "fall back to the full column set" guard would actually be wrong here — it would silently reintroduce this PR's own bug (sorting by an excluded generated column) for that edge case. Added `testVerifyData_withSortOnFilteredColumnsOnlyTrueAndAllColumnsExcluded_doesNotThrow` in 98cda0f5 to lock this in as a regression test rather than just leaving it as reasoning in a comment.

Also fixed CodeRabbit's four inline findings (replied individually on each thread) — one blocking (subclass override hook bypass), two docs/test-coverage suggestions applied, one (MySQL `--` comment syntax) verified against `DdlExecutor.readSqlFromFile()`'s actual line-stripping behavior and declined as not applicable to how this project executes its test DDL.
